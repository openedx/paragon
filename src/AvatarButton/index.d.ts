import React from 'react';

export interface AvatarButtonProps {
  alt?: string,
  size?: 'xs' | 'sm' | 'md',
  src?: string,
}
declare const AvatarButton: React.FC<AvatarButtonProps>;
export default AvatarButton;
