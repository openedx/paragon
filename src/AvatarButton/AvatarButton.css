/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:19 GMT
 */

:root {
  --pgn-spacing-avatar-button-padding-left-lg: .25em;
  --pgn-spacing-avatar-button-padding-left-sm: .25em;
  --pgn-spacing-avatar-button-padding-left-base: .25em;
}
