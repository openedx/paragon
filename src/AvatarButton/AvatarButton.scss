@import "variables";

.btn.pgn__avatar-button-avatar {
  padding-left: $avatar-button-padding-left;

  &.pgn__avatar-button-avatar-sm {
    padding-left: $avatar-button-padding-left-sm;
  }

  &.pgn__avatar-button-avatar-lg {
    padding-left: $avatar-button-padding-left-lg;
  }

  .pgn__avatar {
    margin-top: -.5em;
    margin-bottom: -.5em;
    margin-inline-end: .5em;
  }

  &.pgn__avatar-button-hide-label {
    padding-left: 4px;

    &:not(.dropdown-toggle) {
      padding-right: 4px;
    }

    .pgn__avatar {
      margin-inline-end: 0;
    }
  }
}
