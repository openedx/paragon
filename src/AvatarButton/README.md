---
title: 'AvatarButton'
components:
- AvatarButton
categories:
- Buttonlike
type: 'component'
status: 'New'
designStatus: 'Done'
devStatus: 'Done'
notes: |


---

A button that contains the userâ€™s Avatar.

## Basic Usage

```jsx live
<>
  <div>
    <AvatarButton size="lg" src="https://picsum.photos/128/128/">Casey</AvatarButton>
  </div>
  <div>
    <AvatarButton size="md" src="https://picsum.photos/128/128/">Casey</AvatarButton>
  </div>
  <div>
    <AvatarButton size="sm" src="https://picsum.photos/128/128/">Casey</AvatarButton>
  </div>
</>
```


## With Dropdown

```jsx live
<Dropdown>
  <Dropdown.Toggle as={AvatarButton} src="https://picsum.photos/128/128">
    Casey
  </Dropdown.Toggle>

  <Dropdown.Menu alignRight>
    <Dropdown.Item href="#/action-1">Resume your last course</Dropdown.Item>
    <Dropdown.Item href="#/action-2">Dashboard</Dropdown.Item>
    <Dropdown.Item href="#/action-3">Profile</Dropdown.Item>
    <Dropdown.Item href="#/action-3">Account</Dropdown.Item>
    <Dropdown.Divider />
    <Dropdown.Item href="#/action-3">Sign out</Dropdown.Item>
  </Dropdown.Menu>
</Dropdown>
```

Props set on Dropdown.Toggle get passed through to the "as" component

```jsx live
<Dropdown>
  <Dropdown.Toggle as={AvatarButton} size="sm" src="https://picsum.photos/128/128/">
    Casey
  </Dropdown.Toggle>

  <Dropdown.Menu alignRight>
    <Dropdown.Item href="#/action-1">Resume your last course</Dropdown.Item>
    <Dropdown.Item href="#/action-2">Dashboard</Dropdown.Item>
    <Dropdown.Item href="#/action-3">Profile</Dropdown.Item>
    <Dropdown.Item href="#/action-3">Account</Dropdown.Item>
    <Dropdown.Divider />
    <Dropdown.Item href="#/action-3">Sign out</Dropdown.Item>
  </Dropdown.Menu>
</Dropdown>
```

## No label

For use in mobile viewports or constrained views.

```jsx live
<>
  <div>
    <AvatarButton showLabel={false} size="lg" src="https://picsum.photos/128/128/">Casey</AvatarButton>
  </div>
  <div>
    <AvatarButton showLabel={false} size="md" src="https://picsum.photos/128/128/">Casey</AvatarButton>
  </div>
  <div>
    <AvatarButton showLabel={false} size="sm" src="https://picsum.photos/128/128/">Casey</AvatarButton>
  </div>

  <Dropdown>
    <Dropdown.Toggle  showLabel={false} as={AvatarButton} src="https://picsum.photos/128/128/">
      Casey
    </Dropdown.Toggle>

    <Dropdown.Menu alignRight>
      <Dropdown.Item href="#/action-1">Resume your last course</Dropdown.Item>
      <Dropdown.Item href="#/action-2">Dashboard</Dropdown.Item>
      <Dropdown.Item href="#/action-3">Profile</Dropdown.Item>
      <Dropdown.Item href="#/action-3">Account</Dropdown.Item>
      <Dropdown.Divider />
      <Dropdown.Item href="#/action-3">Sign out</Dropdown.Item>
    </Dropdown.Menu>
  </Dropdown>
</>
```

CSS overrides can be applied to this component through the `.btn-avatar` class name.
