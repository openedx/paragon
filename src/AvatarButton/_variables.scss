$avatar-button-padding-left: .25em !default;
$avatar-button-padding-left-sm: .25em !default;
$avatar-button-padding-left-lg: .25em !default;
