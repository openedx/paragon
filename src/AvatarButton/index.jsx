import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Button from '../Button';
import Avatar from '../Avatar';

const buttonSizesToAvatarSize = {
  sm: 'xs',
  md: 'sm',
  lg: 'md',
};

const AvatarButton = React.forwardRef(({
  children,
  className,
  showLabel,
  size,
  src,
  ...attrs
}, ref) => {
  const avatarSize = buttonSizesToAvatarSize[size] || 'sm';
  return (
    <Button
      {...attrs}
      className={classNames(
        'btn-avatar', // the public class name for custom styling
        'pgn__avatar-button-avatar',
        `pgn__avatar-button-avatar-${size}`,
        className,
        { 'pgn__avatar-button-hide-label': !showLabel },
      )}
      size={size}
      ref={ref}
    >
      <Avatar src={src} alt={showLabel ? '' : children} size={avatarSize} />
      {showLabel && children}
    </Button>
  );
});

AvatarButton.propTypes = {
  /** The button text */
  children: PropTypes.string,
  /** A class name to append to the button */
  className: PropTypes.string,
  /** Show the label or only the avatar */
  showLabel: PropTypes.bool,
  /** The button size */
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  /** Image src of the avatar image */
  src: PropTypes.string,
  /** The button style variant to use */
  variant: PropTypes.string,
};

AvatarButton.defaultProps = {
  children: undefined,
  className: undefined,
  showLabel: true,
  size: 'md',
  src: undefined,
  variant: 'tertiary',
};

export default AvatarButton;
