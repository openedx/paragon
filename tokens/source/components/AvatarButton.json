{
  "avatar-button": {
    "padding": {
      "left": {
        "base": { "value": ".25em", "themeable": true, "source": "$avatar-button-padding-left" },
        "sm": { "value": ".25em", "themeable": true, "source": "$avatar-button-padding-left-sm" },
        "lg": { "value": ".25em", "themeable": true, "source": "$avatar-button-padding-left-lg" }
      }
    }
  }
}
