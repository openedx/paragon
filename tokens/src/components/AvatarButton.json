{
  "spacing": {
    "avatar-button": {
      "padding-left": {
        "base": { "value": ".25em", "source": "$avatar-button-padding-left" },
        "sm": { "value": ".25em", "source": "$avatar-button-padding-left-sm" },
        "lg": { "value": ".25em", "source": "$avatar-button-padding-left-lg" }
      }
    }
  }
}
