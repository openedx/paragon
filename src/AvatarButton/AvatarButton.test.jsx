import React from 'react';
import renderer from 'react-test-renderer';
import AvatarButton from '.';

describe('AvatarButton', () => {
  it('renders in all sizes', () => {
    const tree = renderer.create((
      <>
        <AvatarButton size="sm">Casey</AvatarButton>
        <AvatarButton>Casey</AvatarButton>
        <AvatarButton size="lg">Casey</AvatarButton>
        <AvatarButton showLabel={false} size="sm">Casey</AvatarButton>
        <AvatarButton showLabel={false}>Casey</AvatarButton>
        <AvatarButton showLabel={false} size="lg">Casey</AvatarButton>
      </>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has a default avatar url', () => {
    const testRenderer = renderer.create(<AvatarButton>Casey</AvatarButton>);
    const testInstance = testRenderer.root;
    // test-file-stub is what our fileMock.js returns for all images and svgs
    expect(testInstance.findByType('img').props.src).toBe('test-file-stub');
  });

  it('can set a custom avatar url', () => {
    const profileUrl = 'https://example.com/profile.png';
    const testRenderer = renderer.create(<AvatarButton src={profileUrl}>Casey</AvatarButton>);
    const testInstance = testRenderer.root;
    expect(testInstance.findByType('img').props.src).toBe(profileUrl);
  });
});
