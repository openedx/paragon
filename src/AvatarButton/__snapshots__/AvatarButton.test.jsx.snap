// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarButton renders in all sizes 1`] = `
[
  <button
    className="btn-avatar pgn__avatar-button-avatar pgn__avatar-button-avatar-sm btn btn-tertiary btn-sm"
    disabled={false}
    type="button"
  >
    <img
      alt=""
      className="pgn__avatar pgn__avatar-xs"
      src="test-file-stub"
    />
    Casey
  </button>,
  <button
    className="btn-avatar pgn__avatar-button-avatar pgn__avatar-button-avatar-md btn btn-tertiary btn-md"
    disabled={false}
    type="button"
  >
    <img
      alt=""
      className="pgn__avatar pgn__avatar-sm"
      src="test-file-stub"
    />
    Casey
  </button>,
  <button
    className="btn-avatar pgn__avatar-button-avatar pgn__avatar-button-avatar-lg btn btn-tertiary btn-lg"
    disabled={false}
    type="button"
  >
    <img
      alt=""
      className="pgn__avatar pgn__avatar-md"
      src="test-file-stub"
    />
    Casey
  </button>,
  <button
    className="btn-avatar pgn__avatar-button-avatar pgn__avatar-button-avatar-sm pgn__avatar-button-hide-label btn btn-tertiary btn-sm"
    disabled={false}
    type="button"
  >
    <img
      alt="Casey"
      className="pgn__avatar pgn__avatar-xs"
      src="test-file-stub"
    />
  </button>,
  <button
    className="btn-avatar pgn__avatar-button-avatar pgn__avatar-button-avatar-md pgn__avatar-button-hide-label btn btn-tertiary btn-md"
    disabled={false}
    type="button"
  >
    <img
      alt="Casey"
      className="pgn__avatar pgn__avatar-sm"
      src="test-file-stub"
    />
  </button>,
  <button
    className="btn-avatar pgn__avatar-button-avatar pgn__avatar-button-avatar-lg pgn__avatar-button-hide-label btn btn-tertiary btn-lg"
    disabled={false}
    type="button"
  >
    <img
      alt="Casey"
      className="pgn__avatar pgn__avatar-md"
      src="test-file-stub"
    />
  </button>,
]
`;
